Client/Frontend/Reader/ReaderMode.swift

```diff
@@ -128,23 +140,41 @@ class ReaderMode: BrowserHelper, ReaderModeStyleViewControllerDelegate {
         return "readerModeMessageHandler"
     }
 
-    func userContentController(userContentController: WKUserContentController, didReceiveScriptMessage message: WKScriptMessage) {
-        println("DEBUG: readerModeMessageHandler message: \(message.body)")
-        if let state = ReaderModeState(rawValue: message.body as String) {
-            self.state = state
-            delegate?.readerMode(self, didChangeReaderModeState: state, forBrowser: browser!)
+    private func handleReaderPageEvent(readerPageEvent: ReaderPageEvent) {
+        switch readerPageEvent {
+            case .PageShow:
+                delegate?.readerMode(self, didDisplayReaderizedContentForBrowser: browser!)
         }
     }
 
-    private func constructAboutReaderURL(originalURL: NSURL?) -> NSURL {
-        if let url = originalURL?.absoluteString {
-            if let encoded = url.stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.alphanumericCharacterSet()) {
-                if let aboutReaderURL = NSURL(string: "about:reader?url=\(encoded)") {
-                    return aboutReaderURL
+    private func handleReaderModeStateChange(state: ReaderModeState) {
+        self.state = state
+        delegate?.readerMode(self, didChangeReaderModeState: state, forBrowser: browser!)
+
+        if activateImmediately && state == ReaderModeState.Available {
```

Comment:
This condition seems broken. What if the user clicks something in the reading list that isn't actually readable? For exmaple, pages that are readable on Android but not mobile, pages that require sign-in, clicking a reading list entry while the user is behind a captive portal, etc. Then reader mode won't be enabled, and we'd end up auto-entering reader mode the next time the user _does_ encounter a readable page, which is broken.

One fix would be to remove <code>&& state == ReaderModeState.Available</code> so that <code>activateImmediately</code> will be set to <code>false</code> unconditionally. Note that <code>enableReaderMode()</code> already checks the reader mode state, so checking <code>ReaderModeState.Available</code> here is redundant. Maybe we should rename the method to <code>enableReaderModeIfAvailable()</code>?